backend/de.metas.manufacturing/src/main/java/org/eevolution/productioncandidate/service/produce/PPOrderProducerFromCandidate.java

```diff
@@ -0,0 +1,171 @@
+import org.adempiere.ad.trx.api.ITrxManager;
+import org.eevolution.api.IPPOrderBL;
+import org.eevolution.api.PPOrderCreateRequest;
+import org.eevolution.model.I_PP_Order;
+import org.eevolution.model.I_PP_Order_Candidate;
+import org.eevolution.model.I_PP_Product_Planning;
+import org.eevolution.productioncandidate.async.OrderGenerateResult;
+import org.eevolution.productioncandidate.model.PPOrderCandidateId;
+import org.eevolution.productioncandidate.model.dao.PPOrderCandidateDAO;
+
+import javax.annotation.Nullable;
+import java.util.Comparator;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Stream;
+
+public class PPOrderProducerFromCandidate
+{
+	private final OrderGenerateResult result;
+	private final PPOrderAllocatorService ppOrderAllocatorBuilderService;
+	private final PPOrderCandidateDAO ppOrderCandidatesDAO;
+	private final IPPOrderBL ppOrderService;
+	private final ITrxManager trxManager;
+	private final IProductPlanningDAO productPlanningsRepo;
+	private final Map<ProductPlanningId, I_PP_Product_Planning> planning2RecordShortTimeIndex;
+
+	@Builder
+	public PPOrderProducerFromCandidate(
+			@NonNull final PPOrderAllocatorService ppOrderAllocatorBuilderService,
+			@NonNull final IPPOrderBL ppOrderService,
+			@NonNull final ITrxManager trxManager,
+			@NonNull final PPOrderCandidateDAO ppOrderCandidatesDAO,
+			@NonNull final IProductPlanningDAO productPlanningsRepo)
+	{
+		this.ppOrderAllocatorBuilderService = ppOrderAllocatorBuilderService;
+		this.ppOrderService = ppOrderService;
+		this.trxManager = trxManager;
+		this.ppOrderCandidatesDAO = ppOrderCandidatesDAO;
+		this.productPlanningsRepo = productPlanningsRepo;
+
+		this.result = new OrderGenerateResult();
+		this.planning2RecordShortTimeIndex = new ConcurrentHashMap<>();
+	}
+
+	@NonNull
+	public OrderGenerateResult createOrders(@NonNull final Stream<I_PP_Order_Candidate> candidates, @Nullable final Boolean isDocComplete)
+	{
+		final ImmutableList<PPOrderCandidateToAllocate> sortedCandidates = candidates
+				.filter(orderCandidate -> !orderCandidate.isProcessed())
+				.map(PPOrderCandidateToAllocate::of)
+				.sorted(Comparator.comparing(PPOrderCandidateToAllocate::getHeaderAggregationKey))
+				.collect(ImmutableList.toImmutableList());
+
+		if (sortedCandidates.isEmpty())
+		{
+			return result;
+		}
+
+		processPPOrderCandidates(sortedCandidates, isDocComplete);
+
+		return result;
+	}
+
+	private void processPPOrderCandidates(
+			@NonNull final ImmutableList<PPOrderCandidateToAllocate> ppOrderCandToAllocateSorted,
+			@Nullable final Boolean isDocComplete)
+	{
+		PPOrderAllocator allocator = null;
+		for (final PPOrderCandidateToAllocate ppOrderCandidateToAllocate : ppOrderCandToAllocateSorted)
+		{
+			if (allocator == null)
+			{
+				allocator = ppOrderAllocatorBuilderService.buildAllocator(ppOrderCandidateToAllocate);
+			}
+
+			boolean fullyAllocatedCandidate = false;
+			while (!fullyAllocatedCandidate)
```

Comment:
the logic form here looks a bit weird to me or it's over abstracted....

Why not have a simple logic as we had at PPOrderCandidateAdvisedEventCreator.createMaterialRequests ?

Reply:
also i see a risk to have an infinite loop.

Reply:
well, it's not as simple as it was in `PPOrderCandidateAdvisedEventCreator.createMaterialRequests` mostly because here we can aggregate multiple candidates into a PP_Order and also create multiple PP_Orders from one `PP_Order_Candidate`...

Reply:
understand... but then pls just add some iterations counter here and enforce to not have more than 1000 iterations (for example).
just to make sure we won't run into an infinite loop if allocator.allocate goes funny.